### PR TITLE
feat: Dark theme implementation for Sway Playground

### LINKED ISSUE #45: Dark mode for sway playground!
As a dark mode user myself, I would greatly appreciate it! The website should detect if the user has dark mode set in their browser, and adjust the colors accordingly.

### PR DESCRIPTION
**Pull request by alexprimak58**
Closes #45 

Videos may be a bit laggy as my PC is not quite powerful:)

## Intro
Hey there, I've done a Dark Theme for Sway Playground by creating `themeContext` and then just made other components to know about the current state of `theme`.A bit below I'll explain all my steps in details, just wanted to say that I really want to have this working so I'm open for making changes and enhancing the implementation of the dark theme.

> P.s. As you can see on the videos, in the PC version you have a button to switch between themes + on page load the theme automatically applies depending on your system preferences. On the Phone version, there is no switch button, but it renders the theme that is the same as your system preferences.

## Implementation explanation

I think I'll start with explaining how I set up a `theme` and then I'll go through the components I've edited.

- ### themeContext.ts

Using React context API I created a `context`.It has a theme property (it has 2 states - light or dark), and `setTheme` - to set a theme (on page load or on SwitchThemeButton click).
Also created `useThemeContext` hook so we can make sure theme has been set up properly and with it we will be accessing theme and setTheme.

- ### SwitchThemeButton.tsx

The main functionality just a simple switch theme button, that sets the theme to an opposite one once clicked. 
Other code is styling (sets a Moon Icon when the dark theme is set and a Sun Icon when the light theme is set).

- ### App.tsx

For some reason on github, it shows my changes kinda wrong.
The changes you see below is everything I've done in `App.tsx`:
```
import {Theme,ThemeContext} from './theme/themeContext'  //imported Theme and custom hook
const [theme,setTheme]=useState<Theme>('light'); // Current theme state
```

And wrapped my entire app with Theme Provider so all my components are aware of the current `theme` state: 
```
return (
    <ThemeContext.Provider value={{setTheme,theme}}>
            //all the code in return
    </ThemeContext.Provider>
  );
```
Finally, depending on theme set the background color of our main div:

`background: theme==='light' ? '#F1F1F1': 'black',`

- ### Copyable.tsx

1. Imported custom `useThemeContext` hook with which accessed current `theme` state:

    `const theme = useThemeContext().theme;`

> These steps will be repeated in many places, so instead of just writing these lines, again and again, keep in mind that step as **Step 0**

2. Depending on the theme, apply color to the `div` and `ContentCopyIcon`.

- ### SecondaryButton.tsx

1. Depending on the theme, apply color and bg color to the `Button`.

- ### shared.tsx

1. Created a `BorderColor `function which depending on a theme will change `StyledBorder` color.

2. `DarkThemeStyling` - this is styling for dropdowns and inputs for dark theme,

- ### ExampleDropdown.tsx & ToolchainDropdown.tsx

1. **Imported DarkThemeStyling**

2. Created a `dropdownStyling` which depending on the theme state will use styling from `shared.tsx` or will return nothing (default styling will be applied). And this styling we will pass to `sx` of `FormControl`.

3. Added MenuProps and removed background styling from Select components for styling purposes.

- ### LogView.tsx

1. Depending on the theme, change the color of `LogView`.

- ### SolidityEditor.tsx & SwayEditor.tsx & ComplexParameterInput.tsx

1. Imported another editor theme from AceEditor - `tomorrow_night_bright`

2. With useMemo depending on the theme's current state set `themeStyle` to _tomorrow_night_bright_  |  _chrome_

3. In the AceEditor component assign the theme: `theme={themeStyle}`

- ### DryrunSwitch.tsx

1. Depending on the theme, change the color of `Switch` and `text` close to it.

- ### FunctionCallAccordion.tsx

2. Depending on the theme, change the style of `Accordion`, `ExpandMoreIcon`, and `FormLabel`.

- ### FunctionParameters.tsx

1. I decided to create a custom TableCellComponent to avoid passing the same changes over and over again.

2. Replaced `TableCellComponent` instead of `TableCell`, but in the place of parameters input left `TableCell` to be able to input our parameters.

- ### FunctionToolbar.tsx

1. Depending on the theme, change the color of the title.

- ### InteractionDrawer.tsx

1. Depending on the theme, change the color of the background and text.

- ### ParameterInput.tsx

1. For that I used the same approach as for **ExampleDropdown.tsx & ToolchainDropdown.tsx**

- ### ResponceCard.tsx

1. As usual - just color and background styling depending on the current theme state.

> `background: theme === '' ? 'white' : 'transparent'`
> 
> This line removed just a small white outline when dark theme is set.`

- ### ActionTollbar.tsx

1.Using useEffect hook on page load we check the system preferences of the user and set the theme state to user preferences.

>As I didn't pass dependencies when useEffect should be called: it will be called only once - on page load.

2. `{!isMobile && <SwitchThemeButton />}` - check if we are not in the mobile version then we should be able to see `SwitchThemeButton`.

That is pretty much it.If something needs to be refactored or changed, will be glad to make edits <3

https://github.com/FuelLabs/sway-playground/assets/78500984/8498e600-56f5-43e2-9127-89583e371b06

https://github.com/FuelLabs/sway-playground/assets/78500984/0fa425dd-7a02-4dca-8446-f8b2e8da8385

### CODE REVIEW

File: app/src/App.tsx
Comment (sdankel) [inline app/src/App.tsx:30]: Rather than setting the default to light, we should just call `useContext` here to get the theme context:

```
  const theme = useContext(ThemeContext);
```

File: app/src/App.tsx
Comment (sdankel) [inline app/src/App.tsx:179]: Rather than passing in the value, we can make the Provider check the theme context.

File: app/src/features/toolbar/components/ActionToolbar.tsx
Comment (sdankel) [inline app/src/features/toolbar/components/ActionToolbar.tsx:58]: This should be done in the context provider.

File: app/src/theme/themeContext.ts
Comment (sdankel) [inline app/src/theme/themeContext.ts:1]: Let's move this under `app/src/context/theme.ts`

**Review by sdankel — COMMENTED** [review]
Thank you for the contribution! This is looking really good. I have a few initial comments and I will dig in deeper on Monday.

In Dark Mode, I think all buttons other than compile should have a light gray color for the text rather than green. This highlights the compile button, which is the primary button on the page. It might look better for the compile button to have green text with a dark backrgound. I will experiment with it and test it out soon.

Thank you again!

**Comment by sdankel** [comment]
I noticed also that the theme button disappears when the "interact" drawer is open.

File: app/src/theme/themeContext.ts
Comment (alexprimak58) [reply app/src/theme/themeContext.ts:1]: Done that ✅

File: app/src/App.tsx
Comment (alexprimak58) [reply app/src/App.tsx:30]: I'm afraid with that I won't be able to get & set a theme.

What do you think if I remove in `ActionToolbar.tsx`

```
useEffect(() => {
    if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
      setTheme('dark');
    } else {
      setTheme('light');
    }
  }, []);
```
And make a theme check inside `App.tsx`

```
const [theme, setTheme] = useState<Theme>(window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light');
```

**Comment by alexprimak58** [comment]
Hey there, I'm really glad that you've reviewed my PR and liked my dark theme, thank you! I've made the changes you've commented!

**Comment by alexprimak58** [comment]
> Thank you for the contribution! This is looking really good. I have a few initial comments and I will dig in deeper on Monday.
> 
> In Dark Mode, I think all buttons other than compile should have a light gray color for the text rather than green. This highlights the compile button, which is the primary button on the page. It might look better for the compile button to have green text with a dark backrgound. I will experiment with it and test it out soon.
> 
> Thank you again!

Hey there, I'm really glad that you've reviewed my PR and liked my dark theme, thank you! I've made some changes you've commented on and waiting for your reply regarding dark theme buttons styling.

**Comment by sdankel** [comment]
I like the look in the latest video you posted best (OnHoverGreen.mp4). A couple more comments on the style:

- Could we make the text of the disabled action button text a bit lighter? Darker than the enabled buttons, but light enough so that it's still easily readable.
- Could we match the style of the editor dropdown to light mode? That is, outlined in light grey rather than a green shadow. Like this (But with the arrows grey instead of green):
image

- When hovering over a secondary action button, could we make the background change color to dark green rather than the text color changing? That would mirror what happens in light mode. I think it's better for consistency with other buttons in the app for the background color to change on hover rather than the text color.

Something like this (using the colors from `@fuel-ui/css`):
May-28-2024 18-42-13

File: app/src/features/interact/components/InteractionDrawer.tsx
Comment (sdankel) [inline app/src/features/interact/components/InteractionDrawer.tsx:27]: Rather than having to check `theme === 'light' ?` everywhere, a cleaner solution might be to create a hashmap of the colors used in the app with the mappings for the mode (in theme.ts). For example:

```typescript
interface ColorMapping {
    light: string;
    dark: string;
}

type ColorName = 'gray1' | 'gray2'; // ...names for other colors

const COLORS: Record<ColorName, ColorMapping> = {
    'gray1': {
          'light': darkColors.gray1,
           'dark': lightColors.scalesGreen7,
     }
     // ... all colors used in the app
}

export const useThemeContext = () => {
  {
    const themeContext = useContext(ThemeContext);

    if (!themeContext) {
      throw new Error('Should be wrapped with ThemeContext.Provider!');
    }

    // Convenience function for getting a color based on the current theme
    const themeColor = useCallback((name: ColorName) => COLORS[name][theme], [theme]);

    return {
      setTheme: themeContext.setTheme,
      theme: themeContext.theme,
      themeColor,
    };
  }
};
```

Then in this file you can do: 

```tsx
  const { themeColor } = useThemeContext();
  return (
    <Drawer
      PaperProps={{
        sx: {
          color: themeColor('gray1'),
```

This will also make it easier to ensure that the styles between light and dark mode are "mirrored" everywhere.

**Review by sdankel — COMMENTED** [review]
~You might also want to look into MUI ThemeProvider since the app uses MUI extensively.~

~https://mui.com/material-ui/customization/dark-mode/~

~Here's an example of toggling the mode: https://mui.com/material-ui/customization/dark-mode/#toggling-color-mode~

Actually using `useFuelTheme` is probably better. Here's an example in the documentation site:
https://github.com/FuelLabs/docs-hub/blame/e159d2edc4b4e4494904ac0b96e2b8591a414e97/src/hooks/useTheme.tsx

**Comment by alexprimak58** [comment]
> I like the look in the latest video you posted best (OnHoverGreen.mp4). A couple more comments on the style:
> 
> 
> 
> - Could we make the text of the disabled action button text a bit lighter? Darker than the enabled buttons, but light enough so that it's still easily readable.
> 
> - Could we match the style of the editor dropdown to light mode? That is, outlined in light grey rather than a green shadow. Like this (But with the arrows grey instead of green):
> 
> image
> 
> 
> 
> - When hovering over a secondary action button, could we make the background change color to dark green rather than the text color changing? That would mirror what happens in light mode. I think it's better for consistency with other buttons in the app for the background color to change on hover rather than the text color.
> 
> 
> 
> Something like this (using the colors from `@fuel-ui/css`):
> 
> May-28-2024 18-42-13
> 
> 
> 
> - General note: We should use green sparingly, only where it's used in the current light mode. Where light mode has black, dark mode should have light. Where light mode has dark grey, dark mode should have light grey, etc. The fuel color scheme is mostly black and white with only a few hints of green where emphasis is desired. I believe the only green in the whole app is in the action toolbar buttons & spinner. Not at all in the "Interact" drawer.

Thank you for all the comments/suggestions I've readthem all,will make changes during weekends and come back with an updated version!

File: app/src/features/interact/components/InteractionDrawer.tsx
Comment (alexprimak58) [reply app/src/features/interact/components/InteractionDrawer.tsx:27]: Thank you for suggesting it, this is really an extremely convenient approach, done that.

**Comment by alexprimak58** [comment]
Good day! I think I've made all the changes you mentioned, tried to stick to black/white styling that time but if some colors need to be changed - I'll change them.

**Comment by sdankel** [comment]
This is looking really good! I'll do another review of the code tomorrow. A couple things I noticed from the preview:

It's a bit hard to tell this is a toggle
image

The disabled compile button should be a darkened green
image

Needs a bit more contrast on the log timestamps to be easily readable
image

**Review by sdankel — COMMENTED** [review]
As a heads up, I'm adding linting and formatting to CI.

You can copy the changes from package.json in [this PR](https://github.com/FuelLabs/sway-playground/pull/91) to run the lint + formatter scripts on this code, or wait until that PR is merged and merge from master to run them. There might be merge conflicts but they'll be simple things like changing `'`s to `"`s

`npm run lint` and `npm run format`

**Comment by alexprimak58** [comment]
Thank you for an update, woking on the changes 🫡

File: app/src/features/editor/components/ExampleDropdown.tsx
Comment (alexprimak58) [reply app/src/features/editor/components/ExampleDropdown.tsx:45]: Just to be clear, should I create a styled component in the file they will be used or move it to `shared.tsx`?As for ToolchainDropdown and ExampleDropdown, I have the same styling.

Comment (alexprimak58) [reply app/src/features/editor/components/ExampleDropdown.tsx]: Done the same for ToolchainDropdown.

**Comment by alexprimak58** [comment]
Good day, I hope I didn't miss anything. Now it is looking way better than in my first commit)) I've been thinking: Would it be a good idea to move styled custom components to `shared.tsx` or they are good to go?Is there anything else I should change?

File: app/src/features/editor/hooks/useLog.tsx
Comment (sdankel) [inline app/src/features/editor/hooks/useLog.tsx:36]: This should use `themeColor`, and remain as `darkColors.gray6` in light mode. The dividers aren't currently visible in dark mode

File: app/src/context/theme.ts
Comment (sdankel) [inline app/src/context/theme.ts:30]: nit: this isn't actually a color, it's the name of the editor theme. It might be better to return `editorTheme` from `useTheme` so it's clear that it's distinct / that it shouldn't be used as a CSS color.

```
export default function useTheme() {
// ...
  const editorTheme = useMemo(
    () => currentTheme === "light" ? "chrome" : "tomorrow_night_bright",
    [currentTheme],
  );
  return {
    theme: currentTheme,
    editorTheme, ...
// ...
```

File: app/src/context/theme.ts
Comment (sdankel) [inline app/src/context/theme.ts:63]: nit: I noticed these are all mapping `darkColors.gray6` to a different color in dark mode. Should they all be the same? Perhaps, just keep `gray2`?

**Review by sdankel — COMMENTED** [review]
Thanks for the quick turnaround. I left a couple more nits/questions, then I think it's ready to go 🚀

**Comment by alexprimak58** [comment]
I made some small changes, as you suggested.

Regarding dividers in dark mode, I could create a custom `StyledDivider` in `useLog.tsx` but with that approach, styling doesn't change even if the theme is changed.So I created `StyledDivider` in `shared.tsx`, and it works fine on theme change.What do you think about this approach, should I leave it like that or find another way?

**Comment by alexprimak58** [comment]
Also how about this copy Icon green - I remember you said that green should be used just for the loader and compiler button.Just my suggerstion that it looks better with green there.
image

File: app/src/context/theme.ts
Comment (sdankel) [inline app/src/context/theme.ts:96]:
```suggestion
    () => (currentTheme === "light" ? "chrome" : "tomorrow_night"),
```

**Review by sdankel — COMMENTED** [review]
Sorry, I had a couple more comments from Friday that I forgot to post. I will go ahead and commit them if I can and then approve. Thanks for your hard work on this PR!

**Review by sdankel — APPROVED** [review]

**Comment by alexprimak58** [comment]
> Sorry, I had a couple more comments from Friday that I forgot to post. I will go ahead and commit them if I can and then approve. Thanks for your hard work on this PR!

Thank you so much, it was a please for me!

**Pull request merged**
